hw/top/defs.bzl

```diff
@@ -0,0 +1,54 @@
+# Copyright lowRISC contributors (OpenTitan project).
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+load("//rules/opentitan:hw.bzl", "opentitan_top")
+load("//hw/top_earlgrey/data/autogen:defs.bzl", "EARLGREY")
+load("//hw/top_darjeeling/data/autogen:defs.bzl", "DARJEELING")
+
+ALL_TOPS = [
+    EARLGREY,
+    DARJEELING,
+]
+
+ALL_TOP_NAMES = [
+    top.name
+    for top in ALL_TOPS
+]
+
+def _all_ip_names():
+    names = {ip.name: 1 for top in ALL_TOPS for ip in top.ips}
+    return sorted(names.keys())
+
+ALL_IP_NAMES = _all_ip_names()
+
+def opentitan_if_ip(ip, obj, default):
+    """
+    Return a select expression that evaluate to `obj` if the ip is
+    supported by the active top, and `default` otherwose.
+    """
+    compatible_tops = []
+    for top in ALL_TOPS:
+        for _ip in top.ips:
+            if _ip.name == ip:
+                compatible_tops.append(top.name)
+                break
+
+    return select({
+        "//hw/top:is_{}".format(top): obj
```

Comment:
This top matching may be okay, too, but we could also add constraint values for the IPs directly and define platforms for the tops. Since it's all hidden underneath `opentitan_require_ip()`, I think we can defer that decision as well, though.

Reply:
The decision to hide the details behind macros is a deliberate one so we are not bound by the current implementation and could switch to platforms in the future. We had a long discussion about platforms with @cfrantz and I am not convinced there are the way to go for tops, at least in their current state (there were some improvements in bazel 8.0). It would require one constraint value per IP and top, and in the future we might want more (like "does the top have an embedded flash"). All of this for no obvious benefit over the current implementation with a string flag?

It seems to me that platforms are more useful for toolchain related decisions, and therefore we might want to use them to replace execution environment. Constraint value could be very useful to represent the test platform/board.